lib/puppet/transaction/resource_harness.rb

```diff
@@ -92,6 +93,25 @@ def perform_changes(resource, context)
     end
 
     capture_audit_events(resource, context)
+    persist_system_values(resource, context)
+  end
+
+  # We persist the last known values for the properties of a resource after resource
+  # application.
+  # @param [Puppet::Type] resource resource whose values we are to persist.
+  # @param [ResourceApplicationContent] context the application context to operate on.
+  def persist_system_values(resource, context)
```

Comment:
I was worried about space too. Our original plan was to do this all in PuppetDB, but we only get the catalog which contains the full string, so it would have been hard to get the md5 instead etc. Keeping it in the agent, we benefit from having these short-forms in hand, and its native comparison capability also, which makes the solution accurate (well as accurate as insync?).